lib/auth/auth.go

```diff
@@ -2554,6 +2554,54 @@ func (a *Server) DeleteNetworkRestrictions(ctx context.Context) error {
 	return a.Services.Restrictions.DeleteNetworkRestrictions(ctx)
 }
 
+// UpsertLock upserts a lock and emits a related audit event.
+func (a *Server) UpsertLock(ctx context.Context, lock types.Lock) error {
```

Comment:
Have you considered decorating services.Access with an "AuditableAccess" variant instead of adding methods to auth.Server? This looks like it might be useful in other contexts, plus we could avoiding growing this file further.

Reply:
Although the `AuditableAccess` abstraction sounds like a neat idea, it isn't that straightforward to implement since the `auth.Server` wrappers depend on other auth helpers/services. What I did is to move the `services.Access`-related code to a separate file `lib/auth/access.go` similar to how it's done with `lib/auth/{user,saml,oidc,trustedcluster}.go`.

Reply:
I'm not sure I get your point. For reference, this is what I had in mind:

```go
type AuditableAccess struct {
	Access
	Emitter apievents.Emitter
}

func (aa *AuditableAccess) UpsertLock(ctx context.Context, lock types.Lock) error {
	err := aa.Access.UpsertLock(ctx, lock)
	if err != nil {
		return err
	}
	if err := aa.Emitter.EmitAuditEvent(ctx, &apievents.LockCreate{
		/* ... */
	}); err != nil {
		/* ... */
	}
	return nil
}

// TODO: Same for whatever other functions we want to audit.
```

```diff
--- a/lib/auth/auth.go
+++ b/lib/auth/auth.go
@@ -153,7 +153,7 @@ func NewServer(cfg *InitConfig, opts ...ServerOption) (*Server, error) {
                        Presence:             cfg.Presence,
                        Provisioner:          cfg.Provisioner,
                        Identity:             cfg.Identity,
-                       Access:               cfg.Access,
+                       Access:               &services.AuditableAccess{Access: cfg.Access, Emitter: cfg.Emitter},
                        DynamicAccessExt:     cfg.DynamicAccessExt,
                        ClusterConfiguration: cfg.ClusterConfiguration,
                        Restrictions:         cfg.Restrictions,
```

Reply:
@codingllama Note that the `auth` package depends on `services` so it isn't immediately workable to introduce a (cyclic) dependency the other way around. Now, the methods to be included in `services.AuditableAccess` call at least `auth.ClientUsername` and `auth.ClientImpersonator`. The `DeleteRole` method would in addition rely on the `Identity` and `Trust` services. All in all, the idea of a simple independent "Access-cum-Emitter" service becomes more cumbersome than expected.

Reply:
It still seems to me that this is not difficult to do, but let's say that moving the code to separate files (which you already did) is a good compromise.